cmp/layout/TilingFrame.js

```diff
@@ -0,0 +1,264 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2021 Extremely Heavy Industries Inc.
+ */
+import {hoistCmp, useLocalModel, HoistModel} from '@xh/hoist/core';
+import {bindable, observable, action, computed, makeObservable} from '@xh/hoist/mobx';
+import {frame, box} from '@xh/hoist/cmp/layout';
+import {getLayoutProps, useOnResize} from '@xh/hoist/utils/react';
+import {useEffect} from 'react';
+import {minBy, castArray} from 'lodash';
+import composeRefs from '@seznam/compose-react-refs';
+import PT from 'prop-types';
+
+import './TilingFrame.scss';
+
+/**
+ * This component renders its children as space filling tiles of the equal size.
+ * It will generate a tiling layout that attempts to compromise between limiting empty space,
+ * and keeping the tile width / height ratio as close to the desired ratio as possible.
+ *
+ * Supports a number of optional constraints on tile dimensions that can be used to produce more
+ * stable layouts. These should be used judiciously, however, as each constraint limits the ability
+ * of the TilingFrame to fill its available space.
+ */
+export const [TilingFrame, tilingFrame] = hoistCmp.withFactory({
+    displayName: 'TilingFrame',
+    className: 'xh-tiling-frame',
+
+    render({
+        children,
+        className,
+        desiredRatio = 1,
+        spacing = 0,
+        minTileRatio,
+        maxTileRatio,
+        minTileWidth,
+        maxTileWidth,
+        minTileHeight,
+        maxTileHeight,
+        ...props
+    }, ref) {
+        const localModel = useLocalModel(() => new LocalModel()),
+            layoutProps = getLayoutProps(props),
+            items = castArray(children),
+            count = children.length;
+
+        useEffect(() => localModel.setConfigProps({
+            count,
+            desiredRatio,
+            spacing,
+            minTileRatio,
+            maxTileRatio,
+            minTileWidth,
+            maxTileWidth,
+            minTileHeight,
+            maxTileHeight
+        }));
+
+        ref = composeRefs(
+            ref,
+            useOnResize(dimensions => localModel.setDimensions(dimensions), {debounce: 100})
+        );
+
+        return frame({
+            ref,
+            className,
+            ...layoutProps,
+            items: items.map((item, idx) => {
+                if (!localModel.layout) return null;
```

Comment:
little surprising that you make the child an array of n  nulls, rather than a single null

Reply:
I had the same thought, but it seems we'd always be dealing with an array of nulls regardless: https://github.com/xh/hoist-react/blob/develop/core/elem.js#L45